File: frontend/project-manager-frontend/src/services/noteService.ts
import api from '@/lib/api';
import { NoteData } from '@/types/note';

export const noteService = {
  createNote: async (note: NoteData): Promise<NoteData> => {
    const response = await api.post('/api/notes', note);
    return response.data;
  },

  updateNote: async (id: number, note: Partial<NoteData>): Promise<NoteData> => {
    const response = await api.put(`/api/notes/${id}`, note);
    return response.data;
  },

  deleteNote: async (id: number): Promise<void> => {
    await api.delete(`/api/notes/${id}`);
  },

  getNoteById: async (id: number): Promise<NoteData> => {
    const response = await api.get(`/api/notes/${id}`);
    return response.data;
  },

  getNotesByTaskId: async (taskId: number): Promise<NoteData[]> => {
    const response = await api.get(`/api/tasks/${taskId}/notes`);
    return response.data;
  },

  getNotesByProjectId: async (projectId: number): Promise<NoteData[]> => {
    const response = await api.get(`/api/projects/${projectId}/notes`);
    return response.data;
  }
};

File: note-service/src/main/java/com/klug/noteservice/service/NoteService.java
package com.klug.noteservice.service;

import com.klug.noteservice.dto.NoteDTO;
import com.klug.noteservice.entity.Note;
import com.klug.noteservice.repository.NoteRepository;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class NoteService {

    @Autowired
    private NoteRepository noteRepository;

    @Autowired
    private ModelMapper modelMapper;

    public NoteDTO createNote(NoteDTO noteDTO) {
        Note note = modelMapper.map(noteDTO, Note.class);
        note.setLastUpdated(LocalDateTime.now());
        Note savedNote = noteRepository.save(note);
        return modelMapper.map(savedNote, NoteDTO.class);
    }

    public NoteDTO updateNote(Long id, NoteDTO noteDTO) {
        Note existingNote = noteRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Note not found"));

        // Atualize os campos, mas mantenha o ID original
        modelMapper.map(noteDTO, existingNote);
        existingNote.setId(id);
        existingNote.setLastUpdated(LocalDateTime.now());

        Note updatedNote = noteRepository.save(existingNote);
        return modelMapper.map(updatedNote, NoteDTO.class);
    }

    public void deleteNote(Long id) {
        noteRepository.deleteById(id);
    }

    public NoteDTO getNoteById(Long id) {
        Note note = noteRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Note not found"));
        return modelMapper.map(note, NoteDTO.class);
    }

    public List<NoteDTO> getNotesByTaskId(Long taskId) {
        List<Note> notes = noteRepository.findByTaskId(taskId);
        return notes.stream()
                .map(note -> modelMapper.map(note, NoteDTO.class))
                .collect(Collectors.toList());
    }

    public List<NoteDTO> getNotesByProjectId(Long projectId) {
        List<Note> notes = noteRepository.findByProjectId(projectId);
        return notes.stream()
                .map(note -> modelMapper.map(note, NoteDTO.class))
                .collect(Collectors.toList());
    }
}